### PR TITLE
Strict mode

### PR DESCRIPTION
**Pull request by orangemug**
Added strict mode as defined in README todo.

### CODE REVIEW

**Comment by eleith** [comment]
with this proposal, you have two places you can put 'strict' into. as an optional second argument into `validate()` or as a parameter on the schema itself.

i like adding it on to the schema object, as it goes along with `allownull`, `required` and the other options. however, there is not a way to specify the top level object as strict.

i think that is why you added the second argument to `validate`. 

what if we exposed a new method -> `validateStrict` as a complement to `validate` ? 

we could keep things simple and have `validateStrict` make the entire schema strict, so there would be no need to pass in strict on any further embedded schemas.

thoughts?

**Comment by orangemug** [comment]
I think this works as expected, I've added an additional test for the `strict: true` on the schema object. Also here's my use case for having both options.

Having additional args rather than having a new method makes strict easier to turn on/off in debug.

    schema.validate(obj, {
      strict: (process.env.NODE_ENV === "development")
    });

This means for an API you can throw errors in development, but in production you can just ignore them. I know you could also do this via a if/else but I think this is cleaner... thoughts?

I'm not quite sure about this comment

> however, there is not a way to specify the top level object as strict

I think the test I've add proves that it works, let me know if I'm wrong (I wrote this patch a while ago :)). I haven't really got a use case for this at the moment so happy to remove, however it does keep the API consistent with the other options. 

**Note:** the reason for not going with schema object option alone, was that you couldn't then switch it on/off easily on a per request basis. Which I think is where strict differs from the other options

File: lib/schema.js [hunk 1]
```diff
@@ -78,6 +78,18 @@ Schema.prototype.validate = function(data) {
     }
   }
 
+  if(this.schema.strict || (opts && opts.strict)) {
```
Comment (eleith) [inline lib/schema.js]: thinking more on this, i think the example you bring up is a good one and likely also the main one to focus on. the idea of turning 'strictness' is an all or nothing point of view (all validation is strict, or none of it is)

with that in mind, (and also based on your test cases) how about we remove the check for `this.schema.strict` and have strictness be something that gets passed in as an option to validate only.

it still feels a bit hidden, but having it be an option object, the key name itself at least is clear.

you okay with that?

Comment (orangemug) [inline lib/schema.js]: You're right we currently don't use `schema.strict` so best to remove for now

**Pull request merged**